### ISSUE TITLE
Property 'createElement' in type 'Adapter' is not assignable to the same property in base type 'EnzymeAdapter'.

### ISSUE THREAD

**Issue by mike-lischke** [preactjs/enzyme-adapter-preact-pure#161]
When trying to set up my Jest Preact tests like shown on the [Preact website](https://preactjs.com/guide/v10/unit-testing-with-enzyme/):

```typescript
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-preact-pure";

Enzyme.configure({ adapter: new Adapter() });
```

an error is shown for the adapter parameter:

```bash
(property) adapter: Enzyme.EnzymeAdapter
Type 'Adapter' is not assignable to type 'EnzymeAdapter'.
  The types of 'createElement(...).type' are incompatible between these types.
    Type 'string | ComponentType<any>' is not assignable to type 'string | JSXElementConstructor<any>'.
      Type 'ComponentClass<any, {}>' is not assignable to type 'string | JSXElementConstructor<any>'.
        Type 'ComponentClass<any, {}>' is not assignable to type 'new (props: any) => Component<any, any, any>'.
          Construct signature return types 'Component<any, {}>' and 'Component<any, any, any>' are incompatible.
            The types returned by 'render(...)' are incompatible between these types.
              Type 'ComponentChild' is not assignable to type 'ReactNode'.
                Type 'bigint' is not assignable to type 'ReactNode'.ts(2322)
[index.d.ts(740, 5): ]()The expected type comes from property 'adapter' which is declared here on type '{ adapter: EnzymeAdapter; disableLifecycleMethods?: boolean | undefined; }'
```

When I look at the `Adapter` export I see indeed a difference in the definition of `createElement` (compared to `EnzymeAdaptor `). `EnzymeAdapter` returns a `ReactElement` while enzyme-adapter-preact-pure returns a preact VNode.

Versions used:

preact: 10.7.1
enzyme-adapter-preact-pure: 3.4.0

**Comment by mike-lischke** [comment]
A workaround for the build is to change the typings for Adapter.d.ts: change the return value of `createElement` to `ReactElement`.

**Comment by robertknight** [comment]
> Just ran into a similar (if not the same) issue, which looks slated to be fixed by https://github.com/preactjs/enzyme-adapter-preact-pure/pull/124

It turns out the changes in that PR have already been integrated, so this is a different issue.

@mike-lischke can you confirm which versions of React or Enzyme-related types packages (eg. @types/react, @types/enzyme) are present in your project?

**Comment by huahouye** [comment]
@robertknight 
```
npx preact-cli create typescript my-project
cd my-project
npm run build
# OR
npm run dev
```

can reproduce this isssue.

**Comment by mike-lischke** [comment]
This is the list of typings I have installed in my project (which was a React app but has been converted to use preact, if that is of interest/relevance):

```
        "@testing-library/jest-dom": "5.14.1",
        "@testing-library/preact": "2.0.1",
        "@types/enzyme": "3.10.10",
        "@types/jest": "27.0.3",
        "@types/jsdom": "16.2.13",

```

Because these were not the latest versions, I upgraded and rebuilt everything. The error changed however to:

```
        "@testing-library/jest-dom": "5.16.4",
        "@testing-library/preact": "2.0.1",
        "@types/enzyme": "3.10.12",
        "@types/jest": "27.4.1",
        "@types/jsdom": "16.2.14",
```

```bash
..../node_modules/enzyme-adapter-preact-pure/build-cjs/src/Adapter.d.ts
TypeScript error in ..../node_modules/enzyme-adapter-preact-pure/build-cjs/src/Adapter.d.ts(16,5):
Property 'createElement' in type 'Adapter' is not assignable to the same property in base type 'EnzymeAdapter'.
  Type '(type: TimerHandler, props: Object | null, ...children: ReactElement<any, string | ((props: any) => ReactElement<any, any> | null) | (new (props: any) => Component<any, any, any>)>[]) => VNode<...>' is not assignable to type '(type: TimerHandler, props: Object | null, ...children: ReactElement<any, string | ((props: any) => ReactElement<any, any> | null) | (new (props: any) => Component<any, any, any>)>[]) => ReactElement<...>'.
    Call signature return types 'VNode<any>' and 'ReactElement<any, string | ((props: any) => ReactElement<any, any> | null) | (new (props: any) => Component<any, any, any>)>' are incompatible.
      The types of 'type' are incompatible between these types.
        Type 'string | ComponentClass<any, {}> | FunctionComponent<any>' is not assignable to type 'string | ((props: any) => ReactElement<any, any> | null) | (new (props: any) => Component<any, any, any>)'.
          Type 'ComponentClass<any, {}>' is not assignable to type 'string | ((props: any) => ReactElement<any, any> | null) | (new (props: any) => Component<any, any, any>)'.
            Type 'ComponentClass<any, {}>' is not assignable to type 'new (props: any) => Component<any, any, any>'.
              Construct signature return types 'Component<any, {}>' and 'Component<any, any, any>' are incompatible.
                The types returned by 'render(...)' are incompatible between these types.
                  Type 'ComponentChild' is not assignable to type 'ReactNode'.
                    Type 'bigint' is not assignable to type 'ReactNode'.  TS2416

    14 |     nodeToHostNode(node: RSTNode | string): Node | null;
    15 |     isValidElement(el: any): boolean;
  > 16 |     createElement(type: string | Function, props: Object | null, ...children: ReactElement[]): VNode<any>;
       |     ^
    17 |     elementToNode(el: ReactElement): RSTNode;
    18 |     wrapWithWrappingComponent: (node: ReactElement, options?: ShallowRendererProps | undefined) => {
    19 |         RootFinder: typeof RootFinder;
```

but could be solved with the same workaround as mentioned above.

**Comment by robertknight** [comment]
This appears to be a compatibility issue between Preact and the latest `@types/react` package (v18), which is a dependency of `@types/enzyme`. The issue can be reproduced in this repository by:

1. Removing the existing `node_modules` directory
2. Removing the entry for `@types/react` (v16) in `yarn.lock`
3. Re-running `yarn install`. This will install @types/react v18 since @types/enzyme allows any version
4. Running `yarn build`

A workaround is to pin `@types/react` to the latest v17 release (17.0.44). If using yarn this can be done using the [`resolutions` field](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/):

```diff
diff --git a/package.json b/package.json
index 0d17fee..e19e7dd 100644
--- a/package.json
+++ b/package.json
@@ -29,6 +29,9 @@
     "typescript": "^4.6.3",
     "yalc": "^1.0.0-pre.34"
   },
+  "resolutions": {
+    "@types/react": "17.0.44"
+  },
   "peerDependencies": {
     "enzyme": "^3.11.0",
     "preact": "^10.0.0"
```

**Comment by robertknight** [comment]
The specific change in `@types/react` v18.0.0 which has caused the issue is a change in the definition of `ReactFragment`:

@types/react v17:

```ts
type ReactFragment = {} | Iterable<ReactNode>;
type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null | undefined;
```

@types/react v18:

```ts
type ReactFragment = Iterable<ReactNode>;
type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null | undefined;
```

This causes the `ComponentChild` output of a render function from Preact to no longer be directly assignable to React's `ReactNode` type. This in turn causes Preact's `VNode` to no longer be assignable to a React `ReactElement`, since these two types incorporate the type of render function outputs via their `type` field.

My understanding is that the `{}` type in `ReactFragment` was masking a bunch of other differences between the types.

**Comment by robertknight** [comment]
This is not going to be the final solution for this issue, but https://github.com/preactjs/enzyme-adapter-preact-pure/pull/177 may be a useful interim step. It adds a cast per the comments above. It would be helpful if folks affected could check out that branch and confirm that everything builds in their projects.

**Comment by robertknight** [comment]
I published a v4.0.1 patch release which incorporates the partial solution mentioned in https://github.com/preactjs/enzyme-adapter-preact-pure/issues/161#issuecomment-1092662856. There is still and incompatibility between the Preact VNode and React element types, but depending on how your project is set up this may not be a problem. If in your project TypeScript thinks that JSX expressions create React elements rather than Preact elements (whether that is true at runtime or not), then this should work.

**Comment by mike-lischke** [comment]
@robertknight Thanks for the fix. It works for me now again.
